### ISSUE TITLE
External authentication option with authorization code grant?

### ISSUE THREAD

**Issue by jtourlamain** [IdentityServer/IdentityServer3#2661]
### Question

Can I use an external AuthenticationOption that is using the Authorization code grant? Or do I need to use the openIdconnectAuthentication?
If we need to use the openidconnectAuthentication  mechanism is it possible when the 3rd party doesn't have the well-known configuration?

**Comment by brockallen** [comment]
IdSvr doesn't know or care how you dot he external authentication as long as you have a middleware for it. That's unrelated to what flow the client is using in IdSvr.

**Comment by jtourlamain** [comment]
I know I can use the Microsoft.Owin.Security.OpenIdConnect middleware to start an external oidc authentication, but which lib can I use for an external authorization grant?

**Comment by leastprivilege** [comment]
You can use identitymodel's TokenClient for that

https://github.com/IdentityModel/IdentityModel

**Comment by jtourlamain** [comment]
Ok, but how can I use the TokenClient in my idsrv configuration (cf ConfigureIdentityProviders) so it connects to an external auth server?

```C#
...
app.Map("/identity", idsrvApp =>
            {
                var factory = new IdentityServerServiceFactory()
                    .UseInMemoryUsers(Users.Get())
                    .UseInMemoryClients(Clients.Get())
                    .UseInMemoryScopes(Scopes.Get());

                _userService = new UserService();
                factory.UserService = new Registration<IUserService>(resolver => _userService);
                idsrvApp.UseIdentityServer(new IdentityServerOptions
                {
                    SiteName = "Appius IdentityServer",
                    IssuerUri = IdsrvConstants.IdSrvIssuerUri,
                    Factory = factory,
                    SigningCertificate = LoadCertificate(),
                    AuthenticationOptions = new AuthenticationOptions()
                    {
                        IdentityProviders = ConfigureIdentityProviders
                    }
                });
            }
            );
        }

        private void ConfigureIdentityProviders(IAppBuilder app, string signInAsType)
        {
            // How can I configure to use an external authorization grant?
        }
```

**Comment by leastprivilege** [comment]
Sorry - I don't follow. What are you trying to achieve?

have you read this https://identityserver.github.io/Documentation/docsv2/configuration/identityProviders.html

**Comment by jtourlamain** [comment]
Yes I've read the document. We're already using the OpenIdConnectAuthenticationMiddleware to connect to one of our partners.

We have a mobile app that connects to myIdsrv. There is a button that allows a user to use external authentication for partnerA and one for partnerB.
PartnerA uses also identity server and everything aligns well.

PartnerB however says they are using openidconnect, but they use other options like Response_mode should be"query", response_type should be "code%20id_token",...

The logging on idsrv to look what's wrong (via serilog) doesn't help us a lot ...
The last log we get is: Triggering challenge for external identity provider
When the external authentication is done, we see again the default identity screen with an url like:

https://myidsrv.local/identity?code=25216e4f81da09cb1f20edc9ead3ea90c957ee1b&state=OpenIdConnect.AuthenticationProperties%3DOC8FibWaEMV-eoMbtJp6FPtetaxFbrMYW_oipDhjY8CntI4y676NF-WoXyskEfsmrt4dT73j_RrgBqmafmJvgM5tTY0dj8wWI5IQFjyj4_zuaZz92WMBtmdPZK5OHg5Hy83HxQ40EWA7RdSYzWBI_lxmdwSKFRDo07Kwx_hE7NPi74CCiybdeJiwW34l7Kip5VC4YrvASY6g6x89IfK2xNJ79mcmgzGILPWjyCII__cvcjED2JnaVPklubyDbXzQOQbfKNYIORvxCrxf2xfNDwDkWfYjJ5dbOXZ4lHDqbp8&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJ0ZXN0aXNzdWVyIiwic3ViIjo0NjUsImF1ZCI6InVyYmktdjEiLCJpYXQiOjE0NTcwODAyNzYsImV4cCI6MTQ1NzA4Mzg3NiwiYXV0aF90aW1lIjoxNDU3MDgwMjc2LCJub25jZSI6IjYzNTkyNjc3MDcyNTUxOTQ1MC5OemszTkdaa1ltVXRZalkwWXkwME5HWTBMVGcwTjJFdE1qa3hPVE0xWldWbE16QmxPV0l3TVRJd05tSXROelpsWVMwME5Ea3dMV0kwTTJRdE1qUTJNMkptWVRjNU9HVTQifQ.ZeF5m5ESa09csk92QwFf5o_9FU6wrguw5aqNIEEvROG8bD7beBnVBoYzhyoyiUg2s_epuYaZ7EK6FlXBwSiUqZ8yHmmPMQ2QUth4AyiDOqFt7i6wli9u5fA28_5NAXzUhnvd9BLffuHwCcKSZX-auE5c7RGUhkojvelpBaZwVS8N4MYwh4Qjg_317cJ04X6n20KARpVXNQxdy0Nbcl8AhfDQvRRnHWcCffoXSZDKiMx-FbjNKVnObPM6AS7xfJ482ov6jKGgGbEJdTiavLzZNO-C6JSDWJXGMwc8AvsnvAq3XjRrsrDPVaMf83-Nou6R81uUegipc3DI48O3DVsopw


No other events are triggered on identity server (like SecurityTokenReceived, SecurityTokenValidated,...)
And no redirect to https://myidsrv.local/identity/permissions.
It's like identity server doesn't pick up the GET request.
if code and id_token are posted towards https://myidsrv.local/identity we get an error: The requested resource does not support http method 'POST'.
Should they use https://myidsrv.local/identity/callback as return uri?

Because the two OIDC systems don't align,  they asked if we can't implement an authorization code flow from our Idsrv towards their server. So a user with mobile phone should then see myidsrv where he can click on external PartnerB which should trigger an Authorization code flow. I don't know if the authorization code flow is here a good option as it should only be used by a confidential client. 

Long story short: how can I implement an authorization code flow towards an External Partner from idsrv or customise the openidconnect parameters to grab code and id_token from url?

**Comment by brockallen** [comment]
I would build a stand alone test client app to use this OIDC server. Once you get that all working, mapping it in IdSvr will be easier.

**Comment by jtourlamain** [comment]
That will probably be easier to start with. Once I have the client working, where can I hook it into idsrv? Is it done via custom AuthenticationMiddleware or how can it be done? Is there a get started example?

**Comment by brockallen** [comment]
> Once I have the client working, where can I hook it into idsrv?

The same place all the others are that you said you were already using. Do the stand alone app first to get your head wrapped around what's needed. I _think_ that will bring clarity.

**Comment by jtourlamain** [comment]
If I use an Android WebView and make use of the IdentityModel.Client.AuthorizeRequest I can get the needed tokens, because I can check on the redirectUrl.
When I setup a MVC app and use the OpenIdConnectAuthentication, it seems that the middleware isn't picking up the redirect Uri.

**Comment by jtourlamain** [comment]
A MVC app with custom authentication works (as in the example MVC Codeflowclient manual). But how can I map it in in my own idsrv? 
Now that I can make it work manually with custom MVC controllers, I still need to combine it with a token that I receive from my own idsrv.

**Comment by brockallen** [comment]
Are you saying that you want some of the tokens returned from the external provider to be claims in the token IdSvr issues?

**Comment by jtourlamain** [comment]
yes that's exactly what I need.

**Comment by brockallen** [comment]
Any claims you issue from your authentication middleware will be available in the claims passed into the authenticate external on the user service. You can then issue them as additional claims on the `AuthenticateResult` on your return from the user service. Then those additional claims on the `AuthenticateResult` are available on the `Subject` in your `GetProfile` API on the user service. You can then capture those and add them to the issued claims. I know that's a lot of steps... hopefully that makes sense.

**Comment by jtourlamain** [comment]
I have a custom MVC Controller (in the same project as myidsrv) that gets the external access_token and id_token. I use the Request.GetOwinContext().Authentication.SignIn(id);  
However my UserService AuthenticateExternal is never called.

**Comment by brockallen** [comment]
It's because you didn't authenticate via middleware that's hooked up into IdSvr, like these from our host sample: https://github.com/IdentityServer/IdentityServer3/blob/master/source/Host.Configuration/IdentityServerExtension.cs#L88

**Comment by jtourlamain** [comment]
So, as the UseOpenIdConnectAuthentication option isn't working with the external idsrv, I wrote a custom MVC controller (based on the example MVC Codeflowclient manual). 
If I understand your response well, I do need to write my custom middleware to be able connect idsrv with the manual flow I wrote (like for ex. the TwitterAuthenticationMiddleware)? Then, once that is connected, I can add all the needed claims in my userservice.

**Comment by leastprivilege** [comment]
I really don't understand what your problem is - the oidc MW works fine to connect to idsrv- it does not support all response types - but code id_token works fine.

**Comment by jtourlamain** [comment]
Our idsrv needs to have an external authentication to one of our partners. They don't work with identityserver and they don't have a well-known config. They do however support openid 

I wrote a custom MVC controller that can receive a code id_token from our partner. We also need to add the tokens returned from the external provider to be claims in the token IdSvr issues. In that way the client can make use of our API (with the token received form our idsrv) and also use the API of our partner (with the token from the external authentication that we wrapped in the id_token our idsrv issues). I tried out the concept between two idsrvs and it works.

The problem is that the oidc MW doesn't work or at least I don't get it working with towards our partner. When they added a hard-coded well-known config, the user was redirected after authenticating and concent screen to our idsrv on:

https://myidsrv.local/identity?code=25216e4f81da09cb1f20edc9ead3ea90c957ee1b&state=OpenIdConnect.AuthenticationProperties%3DOC8FibWaEMV-eoMbtJp6FPtetaxFbrMYW_oipDhjY8CntI4y676NF-WoXyskEfsmrt4dT73j_RrgBqmafmJvgM5tTY0dj8wWI5IQFjyj4_zuaZz92WMBtmdPZK5OHg5Hy83HxQ40EWA7RdSYzWBI_lxmdwSKFRDo07Kwx_hE7NPi74CCiybdeJiwW34l7Kip5VC4YrvASY6g6x89IfK2xNJ79mcmgzGILPWjyCII__cvcjED2JnaVPklubyDbXzQOQbfKNYIORvxCrxf2xfNDwDkWfYjJ5dbOXZ4lHDqbp8&id_token=eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJ0ZXN0aXNzdWVyIiwic3ViIjo0NjUsImF1ZCI6InVyYmktdjEiLCJpYXQiOjE0NTcwODAyNzYsImV4cCI6MTQ1NzA4Mzg3NiwiYXV0aF90aW1lIjoxNDU3MDgwMjc2LCJub25jZSI6IjYzNTkyNjc3MDcyNTUxOTQ1MC5OemszTkdaa1ltVXRZalkwWXkwME5HWTBMVGcwTjJFdE1qa3hPVE0xWldWbE16QmxPV0l3TVRJd05tSXROelpsWVMwME5Ea3dMV0kwTTJRdE1qUTJNMkptWVRjNU9HVTQifQ.ZeF5m5ESa09csk92QwFf5o_9FU6wrguw5aqNIEEvROG8bD7beBnVBoYzhyoyiUg2s_epuYaZ7EK6FlXBwSiUqZ8yHmmPMQ2QUth4AyiDOqFt7i6wli9u5fA28_5NAXzUhnvd9BLffuHwCcKSZX-auE5c7RGUhkojvelpBaZwVS8N4MYwh4Qjg_317cJ04X6n20KARpVXNQxdy0Nbcl8AhfDQvRRnHWcCffoXSZDKiMx-FbjNKVnObPM6AS7xfJ482ov6jKGgGbEJdTiavLzZNO-C6JSDWJXGMwc8AvsnvAq3XjRrsrDPVaMf83-Nou6R81uUegipc3DI48O3DVsopw

but it doesn't redirect further.
Log statements are:
- External login requested for provider: "externalpartner"
- Triggering challenge for external identity provider
- Welcome page requested - rendering
and that's it (no log like callback is invoked from external identity provider). 

That's why I started a manual flow via a MVC Controller.

**Comment by leastprivilege** [comment]
Iirc the ms mw requires a form post response and an id token. 

So you probably need your own middleware, or your partner needs a more flexible oidc implementation. You could fork the mw mw as a starting point.

**Comment by jtourlamain** [comment]
Thank you both for replying to all my questions. When combining all the responses I think I can rework my current code to get a complete working solution.

**Comment by naveenkumarreddy** [comment]
Is there any way to limit windows authentication to specific users or AD group?
